dashboard/app/lib/javascripts/dashboard/main.js

```diff
@@ -65,6 +72,115 @@ window.Dashboard = {
 		Marbles.history.navigate('login'+ redirectPath);
 	},
 
+	__catchInsecurePingResponse: function(httpsArgs, resolve) {
+		var httpsXhr = httpsArgs[1], self = this,
+			handleSuccess, handleError, p;
+
+		handleSuccess = function (httpArgs, resolve) {
+			var httpXhr = httpArgs[1];
+			// https did not work but http did...something is wrong with the cert
+			self.Dispatcher.handleAppEvent({
+				name: "HTTPS_CERT_MISSING",
+				status: httpXhr.status
+			});
+			resolve();
+		};
+		handleError = function (httpArgs) {
+			var httpXhr = httpArgs[1];
+			if (httpXhr.status === 0) {
+				// https is failing as well...service is unavailable
+				self.Dispatcher.handleAppEvent({
+					name: "SERVICE_UNAVAILABLE",
+					status: httpXhr.status
+				});
+			}
+			// https did not work but http did without a network error
+			// => missing ssl exception for controller
+			self.Dispatcher.handleAppEvent({
+				name: "HTTPS_CERT_MISSING",
+				status: httpXhr.status
+			});
+			resolve();
+		};
+
+		if (httpsXhr.status === 0) {
+			// https is unavailable, let's see if http works
+			self.client.ping("controller", "http").catch(function(args) {
+				handleError(args, resolve);
+			}).then(function(args){
+				handleSuccess(args, resolve)
+			});
+			return;
+		}
+		// an error code other than 0
+		self.Dispatcher.handleAppEvent({
+			name: "SERVICE_UNAVAILABLE",
+			status: httpsXhr.status
+		});
+
+		resolve();
+	},
+
+	__catchSecurePingResponse: function(args, resolve) {
+		var xhr = args[1];
+		if (xhr.status === 0) {
+			// We were not able to access the controller due to a network error (ssl, timeout)
+			// In order to understand what's happening, we have to switch to http.
+			this.Dispatcher.handleAppEvent({
+				name: "CONTROLLER_UNREACHABLE_FROM_HTTPS",
+				status: xhr.status
+			});
+			resolve();
+			return;
+		}
+
+		// an error code other than 0
+		this.Dispatcher.handleAppEvent({
+			name: "SERVICE_UNAVAILABLE",
+			status: xhr.status
+		});
+		resolve();
+	},
+
+	__successPingResponse: function(args, resolve) {
+		var xhr = args[1];
+		if (xhr.status !== 200) {
+			// If we don't make sure that the response was successful, we might end up in an
+			// infinite loop.
+			resolve();
+			return;
+		}
+		window.location.href = window.location.href.replace("http:", "https:");
+		resolve();
+	},
+
+	__isCertInstalled: function() {
+		var self = this;
+		return new Promise(function(resolve, reject) {
+			if (window.location.protocol === "https:") {
+				self.client.ping("controller", "https").catch(function(args) {
```

Comment:
Actually, I missed that you're using the `resolve`/`reject` of the parent promise.